File: sdks/go/builderengine/streaming.go
package builderengine

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"net/url"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// StreamEventType represents the type of streaming event
type StreamEventType string

const (
	StreamEventConnected          StreamEventType = "connected"
	StreamEventDisconnected       StreamEventType = "disconnected"
	StreamEventError              StreamEventType = "error"
	StreamEventCallStarted        StreamEventType = "call.started"
	StreamEventCallRinging        StreamEventType = "call.ringing"
	StreamEventCallAnswered       StreamEventType = "call.answered"
	StreamEventCallEnded          StreamEventType = "call.ended"
	StreamEventCallFailed         StreamEventType = "call.failed"
	StreamEventAudioStart         StreamEventType = "audio.start"
	StreamEventAudioData          StreamEventType = "audio.data"
	StreamEventAudioEnd           StreamEventType = "audio.end"
	StreamEventTranscriptionPartial StreamEventType = "transcription.partial"
	StreamEventTranscriptionFinal StreamEventType = "transcription.final"
	StreamEventUserSpeechStart    StreamEventType = "user.speech.start"
	StreamEventUserSpeechEnd      StreamEventType = "user.speech.end"
	StreamEventAgentSpeechStart   StreamEventType = "agent.speech.start"
	StreamEventAgentSpeechEnd     StreamEventType = "agent.speech.end"
	StreamEventFunctionCall       StreamEventType = "function.call"
	StreamEventFunctionResult     StreamEventType = "function.result"
	StreamEventDTMFReceived       StreamEventType = "dtmf.received"
)

// StreamEvent represents a streaming event
type StreamEvent struct {
	Type      StreamEventType        `json:"type"`
	Data      map[string]interface{} `json:"data"`
	CallID    string                 `json:"call_id,omitempty"`
	Timestamp string                 `json:"timestamp,omitempty"`
}

// EventHandler is a function that handles streaming events
type EventHandler func(event StreamEvent)

// StreamingClient handles WebSocket streaming connections
type StreamingClient struct {
	apiKey     string
	baseURL    string
	conn       *websocket.Conn
	handlers   map[StreamEventType][]EventHandler
	wildcardHandlers []EventHandler
	mu         sync.RWMutex
	running    bool
	reconnectAttempts int
	maxReconnectAttempts int
	done       chan struct{}
}

// NewStreamingClient creates a new streaming client
func NewStreamingClient(apiKey string, baseURL string) *StreamingClient {
	return &StreamingClient{
		apiKey:              apiKey,
		baseURL:             baseURL,
		handlers:            make(map[StreamEventType][]EventHandler),
		wildcardHandlers:    make([]EventHandler, 0),
		maxReconnectAttempts: 5,
		done:                make(chan struct{}),
	}
}

// On registers an event handler for a specific event type
func (s *StreamingClient) On(eventType StreamEventType, handler EventHandler) *StreamingClient {
	s.mu.Lock()
	defer s.mu.Unlock()

	if eventType == "*" {
		s.wildcardHandlers = append(s.wildcardHandlers, handler)
	} else {
		s.handlers[eventType] = append(s.handlers[eventType], handler)
	}
	return s
}

// OnAll registers a handler for all events
func (s *StreamingClient) OnAll(handler EventHandler) *StreamingClient {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.wildcardHandlers = append(s.wildcardHandlers, handler)
	return s
}

// Off removes an event handler
func (s *StreamingClient) Off(eventType StreamEventType, handler EventHandler) *StreamingClient {
	s.mu.Lock()
	defer s.mu.Unlock()

	// Note: Go doesn't have function equality, so we'd need to use function pointers
	// This is a simplified implementation
	return s
}

// ConnectOptions represents options for connecting
type ConnectOptions struct {
	CallID       string
	SubscribeAll bool
}

// Connect connects to the WebSocket server
func (s *StreamingClient) Connect(opts ConnectOptions) error {
	wsURL := s.baseURL + "/ws/v1/stream"

	params := url.Values{}
	if opts.CallID != "" {
		params.Set("call_id", opts.CallID)
	}
	if opts.SubscribeAll {
		params.Set("subscribe_all", "true")
	}
	if query := params.Encode(); query != "" {
		wsURL += "?" + query
	}

	header := make(map[string][]string)
	header["Authorization"] = []string{"Bearer " + s.apiKey}
	header["Sec-WebSocket-Protocol"] = []string{"builderengine"}

	dialer := websocket.Dialer{
		HandshakeTimeout: 10 * time.Second,
	}

	conn, _, err := dialer.Dial(wsURL, header)
	if err != nil {
		return &WebSocketError{Message: "failed to connect", Err: err}
	}

	s.conn = conn
	s.running = true
	s.reconnectAttempts = 0

	// Dispatch connected event
	s.dispatchEvent(StreamEvent{
		Type: StreamEventConnected,
		Data: map[string]interface{}{"url": wsURL, "call_id": opts.CallID},
	})

	// Start reading messages
	go s.readPump(opts)

	return nil
}

// readPump reads messages from the WebSocket
func (s *StreamingClient) readPump(opts ConnectOptions) {
	defer func() {
		s.conn.Close()
		if s.running {
			s.handleReconnect(opts)
		}
	}()

	for {
		select {
		case <-s.done:
			return
		default:
			_, message, err := s.conn.ReadMessage()
			if err != nil {
				if s.running {
					s.dispatchEvent(StreamEvent{
						Type: StreamEventError,
						Data: map[string]interface{}{"error": err.Error()},
					})
				}
				return
			}

			var event StreamEvent
			if err := json.Unmarshal(message, &event); err != nil {
				continue
			}

			s.dispatchEvent(event)
		}
	}
}

// dispatchEvent dispatches an event to all registered handlers
func (s *StreamingClient) dispatchEvent(event StreamEvent) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	// Call specific handlers
	if handlers, ok := s.handlers[event.Type]; ok {
		for _, handler := range handlers {
			go func(h EventHandler) {
				defer func() {
					if r := recover(); r != nil {
						// Log error but don't crash
					}
				}()
				h(event)
			}(handler)
		}
	}

	// Call wildcard handlers
	for _, handler := range s.wildcardHandlers {
		go func(h EventHandler) {
			defer func() {
				if r := recover(); r != nil {
					// Log error but don't crash
				}
			}()
			h(event)
		}(handler)
	}
}

// handleReconnect handles reconnection after disconnect
func (s *StreamingClient) handleReconnect(opts ConnectOptions) {
	if s.reconnectAttempts >= s.maxReconnectAttempts {
		s.dispatchEvent(StreamEvent{
			Type: StreamEventError,
			Data: map[string]interface{}{"error": "max reconnection attempts reached"},
		})
		return
	}

	s.reconnectAttempts++
	delay := time.Duration(1<<uint(s.reconnectAttempts)) * time.Second

	time.Sleep(delay)

	if err := s.Connect(opts); err != nil {
		s.handleReconnect(opts)
	}
}

// Disconnect disconnects from the WebSocket server
func (s *StreamingClient) Disconnect() {
	s.running = false
	close(s.done)

	if s.conn != nil {
		s.conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseNormalClosure, ""))
		s.conn.Close()
	}

	s.dispatchEvent(StreamEvent{
		Type: StreamEventDisconnected,
		Data: map[string]interface{}{"reason": "client_disconnect"},
	})
}

// Send sends a message to the server
func (s *StreamingClient) Send(message map[string]interface{}) error {
	if s.conn == nil || !s.running {
		return &WebSocketError{Message: "not connected"}
	}

	data, err := json.Marshal(message)
	if err != nil {
		return &WebSocketError{Message: "failed to marshal message", Err: err}
	}

	if err := s.conn.WriteMessage(websocket.TextMessage, data); err != nil {
		return &WebSocketError{Message: "failed to send message", Err: err}
	}

	return nil
}

// Subscribe subscribes to events for a specific call
func (s *StreamingClient) Subscribe(callID string) error {
	return s.Send(map[string]interface{}{
		"action":  "subscribe",
		"call_id": callID,
	})
}

// Unsubscribe unsubscribes from events for a specific call
func (s *StreamingClient) Unsubscribe(callID string) error {
	return s.Send(map[string]interface{}{
		"action":  "unsubscribe",
		"call_id": callID,
	})
}

// SendAudio sends audio data to a call
func (s *StreamingClient) SendAudio(callID string, audioData []byte) error {
	encoded := base64.StdEncoding.EncodeToString(audioData)
	return s.Send(map[string]interface{}{
		"action":  "audio",
		"call_id": callID,
		"data":    encoded,
	})
}

// InjectText injects text for the agent to speak
func (s *StreamingClient) InjectText(callID string, text string, interrupt bool) error {
	return s.Send(map[string]interface{}{
		"action":    "inject",
		"call_id":   callID,
		"text":      text,
		"interrupt": interrupt,
	})
}

// SendDTMF sends DTMF tones
func (s *StreamingClient) SendDTMF(callID string, digits string) error {
	return s.Send(map[string]interface{}{
		"action":  "dtmf",
		"call_id": callID,
		"digits":  digits,
	})
}

// IsConnected returns whether the client is connected
func (s *StreamingClient) IsConnected() bool {
	return s.running && s.conn != nil
}

// WaitForEvent waits for a specific event type with timeout
func (s *StreamingClient) WaitForEvent(eventType StreamEventType, timeout time.Duration) (*StreamEvent, error) {
	eventChan := make(chan StreamEvent, 1)

	handler := func(event StreamEvent) {
		select {
		case eventChan <- event:
		default:
		}
	}

	s.On(eventType, handler)
	defer func() {
		// Remove handler (simplified - would need proper implementation)
	}()

	select {
	case event := <-eventChan:
		return &event, nil
	case <-time.After(timeout):
		return nil, fmt.Errorf("timeout waiting for event: %s", eventType)
	}
}

File: sdks/python/builderengine/streaming.py
"""
Builder Engine Python SDK - Streaming Client

This module provides WebSocket streaming capabilities for real-time events.
"""

from __future__ import annotations

import asyncio
import json
import logging
from typing import Optional, Dict, Any, Callable, AsyncIterator, Union
from dataclasses import dataclass
from enum import Enum

try:
    import websockets
    from websockets.client import WebSocketClientProtocol
    HAS_WEBSOCKETS = True
except ImportError:
    HAS_WEBSOCKETS = False

from builderengine.exceptions import WebSocketError, AuthenticationError

logger = logging.getLogger("builderengine.streaming")


class StreamEventType(str, Enum):
    """Types of streaming events."""
    # Connection events
    CONNECTED = "connected"
    DISCONNECTED = "disconnected"
    ERROR = "error"

    # Call events
    CALL_STARTED = "call.started"
    CALL_RINGING = "call.ringing"
    CALL_ANSWERED = "call.answered"
    CALL_ENDED = "call.ended"
    CALL_FAILED = "call.failed"

    # Audio events
    AUDIO_START = "audio.start"
    AUDIO_DATA = "audio.data"
    AUDIO_END = "audio.end"

    # Transcription events
    TRANSCRIPTION_PARTIAL = "transcription.partial"
    TRANSCRIPTION_FINAL = "transcription.final"

    # Conversation events
    USER_SPEECH_START = "user.speech.start"
    USER_SPEECH_END = "user.speech.end"
    AGENT_SPEECH_START = "agent.speech.start"
    AGENT_SPEECH_END = "agent.speech.end"

    # Function events
    FUNCTION_CALL = "function.call"
    FUNCTION_RESULT = "function.result"

    # DTMF events
    DTMF_RECEIVED = "dtmf.received"


@dataclass
class StreamEvent:
    """A streaming event from the WebSocket connection."""
    type: StreamEventType
    data: Dict[str, Any]
    call_id: Optional[str] = None
    timestamp: Optional[str] = None

    @classmethod
    def from_message(cls, message: Dict[str, Any]) -> "StreamEvent":
        """Create a StreamEvent from a WebSocket message."""
        return cls(
            type=StreamEventType(message.get("type", "error")),
            data=message.get("data", {}),
            call_id=message.get("call_id"),
            timestamp=message.get("timestamp"),
        )


# Type alias for event handlers
EventHandler = Callable[[StreamEvent], None]
AsyncEventHandler = Callable[[StreamEvent], Any]


class StreamingClient:
    """
    WebSocket client for real-time streaming events.

    Provides real-time access to call events, transcriptions,
    and audio streams via WebSocket connections.

    Example:
        >>> client = BuilderEngine(api_key="...")
        >>> streaming = client.streaming()
        >>>
        >>> @streaming.on(StreamEventType.TRANSCRIPTION_FINAL)
        ... async def handle_transcription(event):
        ...     print(f"User said: {event.data['text']}")
        >>>
        >>> async with streaming.connect(call_id="call_abc123"):
        ...     await streaming.listen()
    """

    def __init__(
        self,
        api_key: str,
        base_url: str = "wss://api.builderengine.io",
    ) -> None:
        if not HAS_WEBSOCKETS:
            raise ImportError(
                "websockets package is required for streaming. "
                "Install it with: pip install websockets"
            )

        self._api_key = api_key
        self._base_url = base_url
        self._websocket: Optional[WebSocketClientProtocol] = None
        self._handlers: Dict[StreamEventType, list] = {}
        self._running = False
        self._reconnect_attempts = 0
        self._max_reconnect_attempts = 5
        self._reconnect_delay = 1.0

    def on(
        self,
        event_type: Union[StreamEventType, str],
    ) -> Callable[[AsyncEventHandler], AsyncEventHandler]:
        """
        Decorator to register an event handler.

        Args:
            event_type: Type of event to handle

        Returns:
            Decorator function

        Example:
            >>> @streaming.on(StreamEventType.CALL_STARTED)
            ... async def on_call_started(event):
            ...     print(f"Call started: {event.call_id}")
        """
        if isinstance(event_type, str):
            event_type = StreamEventType(event_type)

        def decorator(handler: AsyncEventHandler) -> AsyncEventHandler:
            if event_type not in self._handlers:
                self._handlers[event_type] = []
            self._handlers[event_type].append(handler)
            return handler

        return decorator

    def add_handler(
        self,
        event_type: Union[StreamEventType, str],
        handler: AsyncEventHandler,
    ) -> None:
        """
        Add an event handler.

        Args:
            event_type: Type of event to handle
            handler: Async function to handle the event
        """
        if isinstance(event_type, str):
            event_type = StreamEventType(event_type)

        if event_type not in self._handlers:
            self._handlers[event_type] = []
        self._handlers[event_type].append(handler)

    def remove_handler(
        self,
        event_type: Union[StreamEventType, str],
        handler: AsyncEventHandler,
    ) -> None:
        """
        Remove an event handler.

        Args:
            event_type: Type of event
            handler: Handler to remove
        """
        if isinstance(event_type, str):
            event_type = StreamEventType(event_type)

        if event_type in self._handlers:
            self._handlers[event_type].remove(handler)

    async def connect(
        self,
        call_id: Optional[str] = None,
        subscribe_all: bool = False,
    ) -> "StreamingClient":
        """
        Connect to the WebSocket server.

        Args:
            call_id: Specific call to subscribe to
            subscribe_all: Subscribe to all events (requires permission)

        Returns:
            Self for chaining
        """
        url = f"{self._base_url}/ws/v1/stream"
        params = []

        if call_id:
            params.append(f"call_id={call_id}")
        if subscribe_all:
            params.append("subscribe_all=true")

        if params:
            url = f"{url}?{'&'.join(params)}"

        headers = {
            "Authorization": f"Bearer {self._api_key}",
        }

        try:
            self._websocket = await websockets.connect(
                url,
                extra_headers=headers,
                ping_interval=30,
                ping_timeout=10,
            )
            self._running = True
            self._reconnect_attempts = 0

            logger.info(f"Connected to WebSocket: {url}")

            # Emit connected event
            await self._dispatch_event(StreamEvent(
                type=StreamEventType.CONNECTED,
                data={"url": url, "call_id": call_id},
            ))

        except Exception as e:
            logger.error(f"WebSocket connection failed: {e}")
            raise WebSocketError(f"Failed to connect: {e}")

        return self

    async def disconnect(self) -> None:
        """Disconnect from the WebSocket server."""
        self._running = False

        if self._websocket:
            await self._websocket.close()
            self._websocket = None

            logger.info("Disconnected from WebSocket")

            await self._dispatch_event(StreamEvent(
                type=StreamEventType.DISCONNECTED,
                data={},
            ))

    async def send(self, message: Dict[str, Any]) -> None:
        """
        Send a message to the server.

        Args:
            message: Message to send
        """
        if not self._websocket:
            raise WebSocketError("Not connected")

        await self._websocket.send(json.dumps(message))
        logger.debug(f"Sent message: {message}")

    async def subscribe(self, call_id: str) -> None:
        """
        Subscribe to events for a specific call.

        Args:
            call_id: Call to subscribe to
        """
        await self.send({
            "action": "subscribe",
            "call_id": call_id,
        })

    async def unsubscribe(self, call_id: str) -> None:
        """
        Unsubscribe from events for a specific call.

        Args:
            call_id: Call to unsubscribe from
        """
        await self.send({
            "action": "unsubscribe",
            "call_id": call_id,
        })

    async def send_audio(self, call_id: str, audio_data: bytes) -> None:
        """
        Send audio data to a call.

        Args:
            call_id: Call to send audio to
            audio_data: Raw audio bytes (PCM 16-bit, 16kHz)
        """
        import base64

        await self.send({
            "action": "audio",
            "call_id": call_id,
            "data": base64.b64encode(audio_data).decode("utf-8"),
        })

    async def inject_text(self, call_id: str, text: str, interrupt: bool = False) -> None:
        """
        Inject text for the agent to speak.

        Args:
            call_id: Call to inject into
            text: Text for the agent to speak
            interrupt: Whether to interrupt current speech
        """
        await self.send({
            "action": "inject",
            "call_id": call_id,
            "text": text,
            "interrupt": interrupt,
        })

    async def listen(self) -> None:
        """
        Start listening for events.

        This method blocks until disconnected.
        """
        if not self._websocket:
            raise WebSocketError("Not connected")

        try:
            async for message in self._websocket:
                if not self._running:
                    break

                try:
                    data = json.loads(message)
                    event = StreamEvent.from_message(data)
                    await self._dispatch_event(event)
                except json.JSONDecodeError as e:
                    logger.warning(f"Invalid JSON message: {e}")
                except Exception as e:
                    logger.error(f"Error processing message: {e}")

        except websockets.ConnectionClosed as e:
            logger.warning(f"WebSocket connection closed: {e}")
            await self._handle_disconnect()
        except Exception as e:
            logger.error(f"WebSocket error: {e}")
            raise WebSocketError(str(e))

    async def events(self) -> AsyncIterator[StreamEvent]:
        """
        Async iterator for events.

        Yields:
            StreamEvent objects

        Example:
            >>> async for event in streaming.events():
            ...     print(f"Event: {event.type}")
        """
        if not self._websocket:
            raise WebSocketError("Not connected")

        try:
            async for message in self._websocket:
                if not self._running:
                    break

                try:
                    data = json.loads(message)
                    yield StreamEvent.from_message(data)
                except json.JSONDecodeError:
                    continue
        except websockets.ConnectionClosed:
            pass

    async def _dispatch_event(self, event: StreamEvent) -> None:
        """Dispatch an event to registered handlers."""
        handlers = self._handlers.get(event.type, [])

        for handler in handlers:
            try:
                result = handler(event)
                if asyncio.iscoroutine(result):
                    await result
            except Exception as e:
                logger.error(f"Error in event handler: {e}")

    async def _handle_disconnect(self) -> None:
        """Handle unexpected disconnection."""
        if self._running and self._reconnect_attempts < self._max_reconnect_attempts:
            self._reconnect_attempts += 1
            delay = self._reconnect_delay * (2 ** (self._reconnect_attempts - 1))

            logger.info(f"Reconnecting in {delay}s (attempt {self._reconnect_attempts})")
            await asyncio.sleep(delay)

            try:
                await self.connect()
            except Exception as e:
                logger.error(f"Reconnection failed: {e}")
                await self._handle_disconnect()

    async def __aenter__(self) -> "StreamingClient":
        """Async context manager entry."""
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        """Async context manager exit."""
        await self.disconnect()


class CallEventHandler:
    """
    Handler for call-specific events.

    Provides a simplified interface for handling events
    for a specific call.

    Example:
        >>> async def handle_call(call_id):
        ...     async with CallEventHandler(streaming, call_id) as handler:
        ...         async for event in handler.events():
        ...             if event.type == StreamEventType.TRANSCRIPTION_FINAL:
        ...                 print(f"User: {event.data['text']}")
    """

    def __init__(
        self,
        streaming: StreamingClient,
        call_id: str,
    ) -> None:
        self._streaming = streaming
        self._call_id = call_id
        self._queue: asyncio.Queue[StreamEvent] = asyncio.Queue()

    async def __aenter__(self) -> "CallEventHandler":
        """Subscribe to call events."""
        await self._streaming.subscribe(self._call_id)

        # Add handler for this call
        @self._streaming.on(StreamEventType.TRANSCRIPTION_FINAL)
        async def handler(event: StreamEvent):
            if event.call_id == self._call_id:
                await self._queue.put(event)

        self._handler = handler
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb) -> None:
        """Unsubscribe from call events."""
        await self._streaming.unsubscribe(self._call_id)

    async def events(self) -> AsyncIterator[StreamEvent]:
        """Iterate over events for this call."""
        while True:
            event = await self._queue.get()
            yield event


class TranscriptionHandler:
    """
    Handler for real-time transcription.

    Buffers partial transcriptions and yields complete utterances.

    Example:
        >>> handler = TranscriptionHandler()
        >>> async for text in handler.transcribe(streaming, call_id):
        ...     print(f"User said: {text}")
    """

    def __init__(self, min_confidence: float = 0.8) -> None:
        self._min_confidence = min_confidence
        self._buffer = ""

    async def transcribe(
        self,
        streaming: StreamingClient,
        call_id: str,
    ) -> AsyncIterator[str]:
        """
        Stream transcriptions for a call.

        Args:
            streaming: StreamingClient instance
            call_id: Call to transcribe

        Yields:
            Complete utterance strings
        """
        await streaming.subscribe(call_id)

        async for event in streaming.events():
            if event.call_id != call_id:
                continue

            if event.type == StreamEventType.TRANSCRIPTION_PARTIAL:
                self._buffer = event.data.get("text", "")

            elif event.type == StreamEventType.TRANSCRIPTION_FINAL:
                text = event.data.get("text", "")
                confidence = event.data.get("confidence", 1.0)

                if confidence >= self._min_confidence and text:
                    yield text
                    self._buffer = ""

            elif event.type == StreamEventType.CALL_ENDED:
                break
